runtime/compiler/compile/J9SymbolReferenceTable.hpp

```diff
@@ -201,6 +201,30 @@ class SymbolReferenceTable : public OMR::SymbolReferenceTableConnector
     */
    TR::SymbolReference * findOrFabricateShadowSymbol(TR_OpaqueClassBlock *containingClass, TR::DataType type, uint32_t offset, bool isVolatile, bool isPrivate, bool isFinal,  const char *name, const char *signature);
 
+   /** \brief
+    *     Returns an array shadow symbol reference fabricated for the field of a flattened array element.
+    *
+    *  \param arrayComponentClass
+    *     The array component class that contains the field.
+    *  \param type
+    *     The data type of the field.
+    *  \param fieldOffset
+    *     The offset of the field.
+    *  \param isVolatile
+    *     Specifies whether the field is volatile.
+    *  \param isPrivate
+    *     Specifies whether the field is private.
+    *  \param isFinal
+    *     Specifies whether the field is final.
+    *  \param fieldName
+    *     The name of the field.
+    *  \param fieldSignature
+    *     The signature of the field.
+    *  \return
+    *     Returns an array shadow symbol reference fabricated for the field of a flattened array element.
+    */
+   TR::SymbolReference * findOrFabricateFlattenedArrayShadowSymbol(TR_OpaqueClassBlock *arrayComponentClass, TR::DataType type, uint32_t fieldOffset, bool isVolatile, bool isPrivate, bool isFinal,  const char *fieldName, const char *fieldSignature);
```

Comment:
> how are you expecting the API to be used for a reference to v.i?

Here are the parameters passed to `findOrFabricateFlattenedArrayShadowSymbol`  for` i` and `v.i`, and the tree looks like. 

```
[     7] O^O VALUE PROPAGATION:  Replacing n15n from acall of <jitLoadFlattenableArrayElement> to aloadi
transformFlattenedArrayElementLoad iloadi fieldSymRef: <Q-VT2-array-shadow>.v.i I fieldEntry[0] fieldName v.i fieldSig I type 3 offset 4
transformFlattenedArrayElementLoad iloadi fieldSymRef: <Q-VT2-array-shadow>.i I fieldEntry[1] fieldName i fieldSig I type 3 offset 8
```

```
n16n      treetop
n15n        newvalue  jitNewValue[#101  helper Method] [flags 0x400 0x0 ] (Identityless sharedMemory )
n60n          loadaddr  VT2[#362  Static] [flags 0x18307 0x0 ]
n67n          iloadi  <Q-VT2-array-shadow>.v.i I[#363  final <Q-VT2-array-shadow>.v.i I] [flags 0x20603 0x200 ]
n66n            aladd (internalPtr sharedMemory )
n11n              ==>aload
n65n              ladd (X>=0 )
n63n                lshl
n61n                  i2l
n12n                    ==>iload
n62n                  iconst 3
n64n                lconst 16
n68n          iloadi  <Q-VT2-array-shadow>.i I[#364  final <Q-VT2-array-shadow>.i I +4] [flags 0x20603 0x200 ]
n66n            ==>aladd
```

> I'm guessing that fieldName will be v.i, the arrayComponentClass will be VT2, 

That matches my understanding too.  

```
arrayClass 0x197d00 name: [L (sig: [QVT2;)
arrayComponentClass 0x197a00 name: VT2 (sig: QVT2;)
```

>and that fieldOffset will be from the start of VT2 rather than from the start of VT1.

That is my understanding too as confirmed on the above tree log.

Reply:
@hzongaro I'm wondering if the above comments address your questions. Meanwhile I've created a draft PR (https://github.com/eclipse-openj9/openj9/pull/14098) on the transformation in VP which will also help answer the question on how the shadow symbols will be created for nested fields.

Reply:
Yes, thank you - you can mark this as resolved.